modes.cpp

```diff
@@ -207,9 +207,18 @@ void CBC_CTS_Encryption::ProcessLastBlock(byte *outString, const byte *inString,
 		// steal from next to last block
 		xorbuf(m_register, inString, BlockSize());
 		m_cipher->ProcessBlock(m_register);
-		inString += BlockSize();
-		length -= BlockSize();
-		memcpy(outString+BlockSize(), m_register, length);
+		if (inString == outString)
+		{
+			length -= BlockSize();
+			memcpy(outString, inString+BlockSize(), length);
+			memcpy(outString+BlockSize(), m_register, length);
+		}

```

Comment:
When `inString == outString`, you have to use `memmove`, not `memcpy`.

Also see [Which one to use - memmove() or memcpy() - when buffers don't overlap?](http://stackoverflow.com/q/1960991/608639), [memcpy of overlapping buffers](http://stackoverflow.com/q/25629736/608639), [Glibc change exposing bugs](http://lwn.net/Articles/414467/) and [Strange sound on mp3 flash website](http://bugzilla.redhat.com/show_bug.cgi?id=638477) and